Source/Tools/LinuxEmulation/LinuxSyscalls/Utils/Threads.cpp

```diff
@@ -12,33 +12,53 @@ namespace FEX::LinuxEmulation::Threads {
     void *Ptr;
     size_t Size;
   };
+
+  struct DeadStackPoolItem {
+    void *Ptr;
+    size_t Size;
+    bool ReadyToBeReaped;
+  };
+
   std::mutex DeadStackPoolMutex{};
   std::mutex LiveStackPoolMutex{};
 
-  static fextl::deque<StackPoolItem> DeadStackPool{};
+  static fextl::deque<DeadStackPoolItem> DeadStackPool{};
   static fextl::deque<StackPoolItem> LiveStackPool{};
 
   void *AllocateStackObject() {
     std::lock_guard lk{DeadStackPoolMutex};
-    if (DeadStackPool.size() == 0) {
-      // Nothing in the pool, just allocate
-      return FEXCore::Allocator::mmap(nullptr, FEX::LinuxEmulation::Threads::STACK_SIZE, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
-    }
-
     // Keep the first item in the stack pool
-    auto Result = DeadStackPool.front().Ptr;
-    DeadStackPool.pop_front();
+    void *Ptr{};
+
+    for (auto it = DeadStackPool.begin(); it != DeadStackPool.end();) {
+      auto Ready = std::atomic_ref<bool>(it->ReadyToBeReaped);
+      bool ReadyToBeReaped = Ready.load();
+      if (Ptr == nullptr && ReadyToBeReaped) {
+        Ptr = it->Ptr;
+        it = DeadStackPool.erase(it);
+        continue;
+      }
 
-    // Erase the rest as a garbage collection step
-    for (auto &Item : DeadStackPool) {
-      FEXCore::Allocator::munmap(Item.Ptr, Item.Size);
+      if (ReadyToBeReaped) {
+        FEXCore::Allocator::munmap(it->Ptr, it->Size);
+        it = DeadStackPool.erase(it);
+        continue;
+      }
+
+      ++it;
     }
-    return Result;
+
+    if (Ptr == nullptr) {
+      Ptr = FEXCore::Allocator::mmap(nullptr, FEX::LinuxEmulation::Threads::STACK_SIZE, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
+    }
+
+    return Ptr;
   }
 
-  void AddStackToDeadPool(void *Ptr) {
+  bool *AddStackToDeadPool(void *Ptr) {
```

Comment:
Returning a C pointer to some boolean in a nontrivial (and undocumented) internal data structure is a massive footgun compared to what you're actually trying to do here. What's the reason for doing it like this?

Reply:
We could return a struct with a named member that does the same thing. Ensuring it is smaller than 128-bits to ensure it gets placed in to registers X0/X1.
We can't guarantee that another thread will receive a SIGCHLD to cleanup the thread's stack, so that's out.
We could have the Threads API add the stack to the thread pool and also do the exit syscall itself to keep the implementation details internal.

We really can't do much when the thread's stack is dead.

Reply:
Yes, we don't control the function signature of clone2, while we do with clone3 (until glibc adds a wrapper anyway) Which is why the code for the entrypoint needs to differ.

We need to have the exit syscall usage right next to whatever takes away ownership of the stack from the thread. Since returning from a function that takes ownership away means using a stack which the thread no longer owns.

Reply:
Fair enough. Then let's move everything into `Threads` - an interface that deals performs a process exit still fits reasonably well into that module while keeping the implementation simpler.

Reply:
Implementation moved over to Threads